File: wear/src/main/java/com/pdyjak/powerampwear/music_browser/files/FileItem.java
package com.pdyjak.powerampwear.music_browser.files;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.pdyjak.powerampwear.music_browser.Clickable;
import com.pdyjak.powerampwear.music_browser.MusicLibraryNavigator;
import com.pdyjak.powerampwearcommon.responses.Parent;

public class FileItem implements Clickable {
    @NonNull
    private final MusicLibraryNavigator mMusicLibraryNavigator;

    @Nullable
    public final Parent parent;
    @NonNull
    public final String trackId;
    @Nullable
    final String title;
    @Nullable
    final String artist;
    @Nullable
    final String album;
    final long duration;

    FileItem(@NonNull MusicLibraryNavigator helper, @Nullable Parent parent,
            @NonNull String trackId, @Nullable String title, @Nullable String artist,
            @Nullable String album, long duration) {
        mMusicLibraryNavigator = helper;
        this.parent = parent;
        this.trackId = trackId;
        this.title = title;
        this.artist = artist;
        this.album = album;
        this.duration = duration;
    }

    @Override
    public void onClicked() {
        mMusicLibraryNavigator.selectFile(this, false);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        FileItem fileItem = (FileItem) o;

        if (duration != fileItem.duration) return false;
        if (parent != null ? !parent.equals(fileItem.parent) : fileItem.parent != null)
            return false;
        if (!trackId.equals(fileItem.trackId)) return false;
        if (title != null ? !title.equals(fileItem.title) : fileItem.title != null) return false;
        if (artist != null ? !artist.equals(fileItem.artist) : fileItem.artist != null)
            return false;
        return album != null ? album.equals(fileItem.album) : fileItem.album == null;

    }

    @Override
    public int hashCode() {
        int result = parent != null ? parent.hashCode() : 0;
        result = 31 * result + trackId.hashCode();
        result = 31 * result + (title != null ? title.hashCode() : 0);
        result = 31 * result + (artist != null ? artist.hashCode() : 0);
        result = 31 * result + (album != null ? album.hashCode() : 0);
        result = 31 * result + (int) (duration ^ (duration >>> 32));
        return result;
    }
}

File: wear/src/main/java/com/pdyjak/powerampwear/music_browser/files/FileItem.kt
package com.pdyjak.powerampwear.music_browser.files

import com.pdyjak.powerampwear.music_browser.Clickable
import com.pdyjak.powerampwear.music_browser.MusicLibraryNavigator
import com.pdyjak.powerampwearcommon.responses.Parent

class FileItem(private val mMusicLibraryNavigator: MusicLibraryNavigator, val parent: Parent?,
               val trackId: String, internal val title: String?, internal val artist: String?,
               internal val album: String?, internal val duration: Long, val contextualId: String?)
    : Clickable {

    override fun onClicked() {
        mMusicLibraryNavigator.selectFile(this, false)
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other === null || javaClass != other.javaClass) return false

        val fileItem = other as FileItem

        if (parent != fileItem.parent) return false
        if (trackId != fileItem.trackId) return false
        if (title != fileItem.title) return false
        if (artist != fileItem.artist) return false
        if (album != fileItem.album) return false
        return duration == fileItem.duration
    }

    override fun hashCode(): Int {
        var result = parent?.hashCode() ?: 0
        result = 31 * result + trackId.hashCode()
        result = 31 * result + (title?.hashCode() ?: 0)
        result = 31 * result + (artist?.hashCode() ?: 0)
        result = 31 * result + (album?.hashCode() ?: 0)
        result = 31 * result + (duration xor duration.ushr(32)).toInt()
        return result
    }
}
